Make JBMC commandline 'java' compatible [TG-9888]

peterschrammel opened the pull request:
As a user, I'd like to use `jbmc` in a way similar to `java`.
For example, I'd like to run `java org.cprover.MyApp` resp. `jbmc org.cprover.Myapp`.
Similarly, I'd like to run `java -jar myapp.jar` resp. `jbmc -jar myapp.jar`.
This PR brings `jbmc`, `janalyzer` and `jdiff` a step closer to that: the most common combinations of `-jar` and `-cp` are now supported.

Fixes #2864.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [X] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

peterschrammel committed:
Check whether package.class matches loaded class

Package of specified main class must match loaded
main class.

peterschrammel committed:
Add test for main method in inner class

Java can run with a main method inside an inner class.
We need a test for this case.

peterschrammel committed:
Emit warning about non-existing classpath elements

Java doesn't do that, but it is useful for debugging
wrong classpaths.

peterschrammel committed:
Option for passing goto-binary to JBMC

Similar to -jar, directly passing a file as an
argument is ambiguous and can be mistaken as a class name.

peterschrammel committed:
Show parse tree for given main class in JBMC

We do this for the given main class now,
which lifts the requirement for providing a single class
file only (we don't accept files anymore).

peterschrammel committed:
Set message handler earlier

Registering the classpath elements happens before parse(),
so we would miss any output produced there.

Waiting for the languaget-is-not-a-messaget PR to
resolve that properly.

peterschrammel committed:
Add -jar option

similar to the 'java' executable which accepts
only -jar jarfile.jar and not jarfile.jar.

peterschrammel committed:
Update help text to accept class name as direct argument

We will not accept class files anymore, but only class names.

peterschrammel committed:
Move classpath help text to more prominent place

as they are referenced in the usage text right above.

peterschrammel committed:
Update JBMC user manual

JBMC now accepts class name as direct argument now.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 1d35ecdea256f82c332f2351f027ddb851ada3d3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114658916
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3898da02ea8cf6eab64b9dce5c439c7373768c4b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114722731

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: c9e6b0c97bc86c06643a5dc79e8b5b9418db734d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114875345
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 916aea1a98961ed214da52e3d15e9dacf7a45774).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114875958

peterschrammel commented:
TG is passing.

smowton commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -134,7 +134,35 @@ void lazy_goto_modelt::initialize(
+    // We assume it's Java.
+    const std::string filename = "";
+    language_filet &lf = add_language_file(filename);
+    lf.language = get_language_from_mode(ID_java);
```
Perhaps this option would be better taken as a constructor arg?

smowton commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -134,7 +134,35 @@ void lazy_goto_modelt::initialize(
+    msg.status() << "Parsing ..." << messaget::eom;
+
+    std::ifstream infile(filename);
+    // Both arguments to parse() are discarded.
```
This is specific to Java I suppose -- perhaps we should give `languaget` an alternative init method to be used when it doesn't represent a single file?

peterschrammel replied:
Yes, when no file is given then we assume that it is Java. There's no file extension we could look at to determine the language.

smowton replied:
I mean shouldn't the driver program tell the lazy-goto-model what it's expecting, and then call some other `languaget` entry-point instead of `parse` with a fake parameter?

peterschrammel replied:
> call some other languaget entry-point

I didn't want to open that can of worms... but since I have already started that work there's some hope that things change to the better... that will be a PR with a lot of noise through all language frontends...

thk123 commented on regression/test.pl:
```diff
@@ -33,7 +33,12 @@ sub with_color {
 sub run($$$$$) {
   my ($name, $input, $cmd, $options, $output) = @_;
-  my $cmdline = "$cmd $options '$input' >'$output' 2>&1";
+  my $cmdline;
```
:pick: Typo in commit message for commit `Do not pass empty commandline argument in test.pl `: 
esample -> example

thk123 commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -1029,6 +1029,13 @@ void jbmc_parse_optionst::help()
     "                              manifest (checked in this order), the behavior is\n" // NOLINT(*)
     "                              the same as running jbmc on the corresponding\n" // NOLINT(*)
     "                              class file."
+    " jbmc -jar jarfile            JAR file to be checked\n"
```
:pick: As per guidelines, these help strings should be put in a macro to avoid copy pasting the same info :rabbit:

peterschrammel replied:
Fully agreed. If it's ok to omit the program name from the help text then this seems be doable.

thk123 commented on jbmc/src/janalyzer/janalyzer_parse_options.cpp:
```diff
@@ -694,6 +694,13 @@ void janalyzer_parse_optionst::help()
     "                              manifest (checked in this order), the behavior is\n" // NOLINT(*)
     "                              the same as running janalyzer on the corresponding\n" // NOLINT(*)
     "                              class file."
+    " janalyzer -jar jarfile       JAR file to be checked\n"
```
:rabbit:

thk123 commented on jbmc/src/janalyzer/janalyzer_parse_options.cpp:
```diff
@@ -700,6 +700,12 @@ void janalyzer_parse_optionst::help()
     "                              of the class specified by --main-class or the main\n" // NOLINT(*)
     "                              class specified in the JAR manifest\n"
     "                              (checked in this order).\n"
+    " janalyzer --gb goto-binary   goto-binary file to be checked\n"
```
:rabbit2:

thk123 commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -1035,6 +1035,12 @@ void jbmc_parse_optionst::help()
     "                              of the class specified by --main-class or the main\n" // NOLINT(*)
     "                              class specified in the JAR manifest\n"
     "                              (checked in this order).\n"
+    " jbmc --gb goto-binary        goto-binary file to be checked\n"
```
:rabbit2:

thk123 commented on jbmc/regression/janalyzer-taint/taint-aliasing1/test.desc:
```diff
@@ -1,5 +1,5 @@
 CORE
-aliasing1.class
+aliasing1
```
:no_entry_sign: The commit message does not match with this change
` Make janalyzer use lazy goto model initialization`

thk123 commented on jbmc/regression/janalyzer/too-many-args/test.desc:
```diff
@@ -1,6 +1,6 @@
-^Only one \.class, \.jar or \.gbf file should be directly specified on the command-line
+A
+B
+Please give exactly one class name, and/or use -jar jarfile or --gb goto-binary
```
I'd put this in a different commit to ease testing

thk123 commented on jbmc/regression/jbmc/classpath-class-incorrect-package/test.desc:
```diff
@@ -1,11 +1,9 @@
-KNOWNBUG
-Test.class
+CORE
```
:no_entry_sign: This absolutely should be in its own commit

peterschrammel replied:
This could go into a separate commit. 
Unfortunately all changes to the other tests need to go into the same commit together with changing the interface (including the one that is currently still split out).

thk123 commented on jbmc/regression/jbmc/classpath-class-with-jar/test.desc:
```diff
@@ -1,7 +1,8 @@
+Test
 --classpath `../../../../scripts/format_classpath.sh src.jar .`
 ^EXIT=10$
+^EXIT=10$
```
Changing the tests at the same time as changing how they are opened makes this much harder to review

peterschrammel replied:
That's a superfluous change, actually.

thk123 reviewed (commented):
I got to the misnamed? commit `Make janalyzer use lazy goto model initialization` - please split up the changes to tests more logically to make reviewing easier (i.e. one commit where you just remove `.class` suffix, one commit where you add `EXIT=10` regex so can just be scanned through

tautschnig commented on jbmc/regression/jbmc/classpath-invalid/test-jar.desc:
```diff
@@ -1,10 +1,9 @@
 --
 --
-symex-driven lazy loading should emit "the program has no entry point" but currently doesn't
+This is not quite 'java' behavior as 'java' always includes . in the classpath.
```
Can't we just fix this?

tautschnig commented on jbmc/regression/jbmc/enum_switch/test.desc:
```diff
@@ -1,5 +1,5 @@
 CORE symex-driven-lazy-loading-expected-failure
-com/diffblue/regression/Foo.class
+com/diffblue/regression/Foo
```
Why does this one still use `/`  as a separator?

tautschnig commented on jbmc/regression/jbmc/enum_values_clone/test.desc:
```diff
@@ -1,5 +1,5 @@
 CORE symex-driven-lazy-loading-expected-failure
-com/diffblue/regression/EnumIter.class
+com/diffblue/regression/EnumIter
```
`.` instead of `/`?

tautschnig commented on jbmc/regression/jbmc/enum_values_clone_name/test.desc:
```diff
@@ -1,5 +1,5 @@
 CORE symex-driven-lazy-loading-expected-failure
-com/diffblue/regression/EnumIter.class
+com/diffblue/regression/EnumIter
```
`.` instead of `/`?

tautschnig commented on jbmc/regression/jbmc/internal1/test.desc:
```diff
@@ -1,5 +1,5 @@
 CORE
-com/diffblue/javatest/nestedobjects/subpackage/Order.class
+com/diffblue/javatest/nestedobjects/subpackage/Order
```
`.` instead of `/`?

tautschnig commented on jbmc/regression/jbmc/overlay-class/correct-test.desc:
```diff
@@ -1,13 +1,13 @@
 ^Field definition for java::Test\.x already loaded from overlay class$
 ^Adding symbol from overlay class:  method 'java::Test\.main:\(\[Ljava/lang/String;\)V'$
-^Method java::Test\.<init>:\(\)V exists in an overlay class without being marked as an overlay and also exists in the underlying class
+^Method java::Test\.<init>:\(\)V exists in an overlay class without being marked as an overlay and also exists in the underlying
```
Wrong auto-replacement?

tautschnig commented on jbmc/regression/jbmc/overlay-class/duplicate-test.desc:
```diff
@@ -8,7 +8,7 @@ Test.class
 ^Field definition for java::Test\.x already loaded from overlay class$
 ^Adding symbol from overlay class:  method 'java::Test\.main:\(\[Ljava/lang/String;\)V'$
-^Method java::Test\.<init>:\(\)V exists in an overlay class without being marked as an overlay and also exists in the underlying class
+^Method java::Test\.<init>:\(\)V exists in an overlay class without being marked as an overlay and also exists in the underlying
```
Wrong auto-replacement?

tautschnig commented on jbmc/regression/jbmc/overlay-class/misordered-test.desc:
```diff
@@ -12,5 +12,5 @@ Test.class
-^Adding symbol from overlay class
-exists in an overlay class without being marked as an overlay and also exists in the underlying class
+^Adding symbol from overlay
+exists in an overlay class without being marked as an overlay and also exists in the underlying
```
Wrong auto-replacement?

tautschnig commented on jbmc/regression/jbmc/overlay-class/unmarked-test.desc:
```diff
@@ -12,5 +12,5 @@ Test.class
-^Adding symbol from overlay class
-exists in an overlay class without being marked as an overlay and also exists in the underlying class
+^Adding symbol from overlay
+exists in an overlay class without being marked as an overlay and also exists in the underlying
```
Wrong auto-replacement?

tautschnig commented on jbmc/regression/jbmc/provide_object_implementation/test.desc:
```diff
@@ -1,5 +1,5 @@
 CORE symex-driven-lazy-loading-expected-failure
-java/lang/Object.class
+java/lang/Object
```
`.` instead of `/`?

tautschnig commented on jbmc/src/janalyzer/janalyzer_parse_options.cpp:
```diff
@@ -358,24 +362,77 @@ int janalyzer_parse_optionst::doit()
 
-  if(cmdline.args.size() > 1)
+  if(!((cmdline.isset("jar") && cmdline.args.size() == 0) ||
+       (cmdline.isset("gb") && cmdline.args.size() == 0) ||
```
I'd prefer `.empty()` over `.size() == 0`

tautschnig commented on jbmc/src/janalyzer/janalyzer_parse_options.cpp:
```diff
@@ -358,24 +362,77 @@ int janalyzer_parse_optionst::doit()
   register_languages();
 
-  if(cmdline.args.size() > 1)
+  if(!((cmdline.isset("jar") && cmdline.args.size() == 0) ||
```
Could we push the negation inwards?

peterschrammel replied:
Hm, I actually prefer `not(or_i good_case_i)` to `and_i not(good_case_i)`

tautschnig replied:
I raised this, because the reader needs to 1) figure out matching parentheses and 2) read all the lines at once. Maybe you just want to introduce some named subexpressions?

tautschnig commented on jbmc/src/janalyzer/janalyzer_parse_options.cpp:
```diff
@@ -358,24 +362,77 @@ int janalyzer_parse_optionst::doit()
 
-  if(process_goto_program(options))
+    const std::string filename = "";
+    std::ifstream infile(filename);
```
What is this about - opening a file without a name will surely fail?

peterschrammel replied:
It doesn't fail, just gives a bad stream. It's ugly, though. I'll have to extend `languaget` to avoid this.

tautschnig commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -472,44 +472,49 @@ int jbmc_parse_optionst::doit()
 
-  if(cmdline.isset("show-parse-tree"))
+  if(!((cmdline.isset("jar") && cmdline.args.size() == 0) ||
+       (cmdline.isset("gb") && cmdline.args.size() == 0) ||
```
As above: push negation inwards, use `.empty()`

tautschnig commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -472,44 +472,49 @@ int jbmc_parse_optionst::doit()
+    log.status() << "Parsing..." << messaget::eom;
 
+    const std::string filename = "";
+    std::ifstream infile(filename);
```
As above - seems invalid.

tautschnig commented on jbmc/src/jbmc/jbmc_parse_options.h:
```diff
@@ -50,7 +50,10 @@ class optionst;
+  "(classpath):(cp):" \
+  "(jar):" \
+  "(main-class):" \
+  "(gb):" \
```
In line with @thk123's comment: this repeats, should be factored out.

tautschnig commented on regression/test.pl:
```diff
@@ -33,7 +33,12 @@ sub with_color {
+    $cmdline = "$cmd $options '$input' >'$output' 2>&1";
+  } else {
+    $cmdline = "$cmd $options >'$output' 2>&1";
+  }
```
Does this actually address #3981?

peterschrammel replied:
Yes, it does.

tautschnig replied:
Cool! Please add the "Fixes: #3981" info to the commit message to have it auto-closed.

peterschrammel committed:
Make janalyzer use lazy goto model initialization

This should be harmonized with JBMC to remove some
code duplication at some point,
but there is some difference in the goto passes that are
currently run; thus, just preserving the current behavior.

peterschrammel committed:
Add no-args parse() function to java_bytecode_languaget

We can parse files from the classpath without
any file given as commandline argument.

peterschrammel committed:
Make JBMC and janalyzer take a class name argument

as well as
-jar jarfile
and
-gb gotobinary

peterschrammel committed:
Show parse tree for given main class in janalyzer

To match behaviour of JBMC.

peterschrammel committed:
Remove obsolete test

This test doesn't make any sense because we don't accept
class files anymore.

peterschrammel committed:
Activate and fix some classpath tests

These work with the new JBMC CLI.

peterschrammel committed:
Fix wrong main method in test

Test even failed when class is loaded
because there is no valid main method.
Now it only fails when the class is not loaded.

kroening commented:
`lazy_goto_modelt` is now really Java only, and should not live in goto-programs any more.

peterschrammel commented:
> lazy_goto_modelt is now really Java only, and should not live in goto-programs any more.

@kroening, a solution that could work is to 
1. introduce in `languaget`
```
  virtual bool parse() = 0;
  virtual bool parse(const std::string &path) = 0;
```
to avoid having to supply dummy arguments.
2. deprecate
```
  virtual bool parse(std::istream &instream, const std::string &path)=0;
```
3. give (lazy_)goto_modelt initialize a default language to use with `parse()` (or use the language registry for that, e.g. 'the first language is the default' or similar). The implementation of `parse()` can then deal with the case where no input file is supplied and the language cannot handle that situation (which would then work transparently for C vs Java).

kroening commented:
@peterschrammel This suffers from the following problem: `languaget` was intended to offer an interface that is common to multiple languages, i.e., really the interface that's the intersection.
The plan here is to make it the union, which isn't good practice.
I would suggest the following fix:
1) keep languaget as is, for the time being (that may be revisited, but seems independent of this PR)
2) jbmc is Java specific, as its name suggests; make jbmc talk to the Java frontend directly. If you like, introduce an interface for that to separate implementation from interface, but for now there would only be one provider.

The bottom line is that the idea to have a common language interface isn't helping if cbmc talks to C, and jbmc talks to Java. The code gets more complicated, not easier.

peterschrammel committed:
Fix include order

peterschrammel committed:
Output final classpath on debug level

romainbrenguier commented on src/goto-programs/initialize_goto_model.cpp:
```diff
@@ -151,11 +152,21 @@ goto_modelt initialize_goto_model(
+      goto_model.symbol_table, options, message_handler);
+
+    // To create a new entry point we must first remove the old one
+    remove_existing_entry_point(goto_model.symbol_table);
```
:bulb: would it have been possible to directly use the right entry point?

peterschrammel replied:
The entry point is part of a goto-binary we loaded.

romainbrenguier commented on src/goto-programs/initialize_goto_model.cpp:
```diff
@@ -151,11 +152,21 @@ goto_modelt initialize_goto_model(
+
+    // Remove the function from the goto functions so it is copied back in
+    // from the symbol table during goto_convert
+    if(!entry_point_generation_failed)
```
:question: is it safe to continue when it failed?

peterschrammel replied:
It will throw a `invalid_source_file_exceptiont` in line 191

romainbrenguier commented on src/goto-programs/rebuild_goto_start_function.cpp:
```diff
@@ -30,8 +30,9 @@ std::unique_ptr<languaget> get_entry_point_language(
+  // Get the relevant languaget to generate the new entry point with.
   std::unique_ptr<languaget> language = get_language_from_mode(mode);
+  // This might fail if the driver program hasn't registered that language.
   INVARIANT(language, "No language found for mode: " + id2string(mode));
```
:pick: maybe rephrase the invariant message so that is phrase as a "should" statement and includes as much  information as the comment (which can then be removed)

romainbrenguier commented on jbmc/src/janalyzer/janalyzer_parse_options.cpp:
```diff
@@ -701,6 +701,21 @@ void janalyzer_parse_optionst::help()
+    #ifdef _WIN32
+    ";"
+    #else
+    ":"
```
:pick: this could be put in a `java_file_separator_char` function, as it can be used in several places of jbmc

romainbrenguier commented on jbmc/src/jbmc/jbmc_parse_options.h:
```diff
@@ -26,13 +26,13 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <solvers/strings/string_refinement.h>
 
 #include <java_bytecode/java_bytecode_language.h>
+#include <java_bytecode/lazy_goto_model.h>
```
The "Move lazy_goto_modelt to JBMC" commit doesn't seem to match its description

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -328,6 +315,45 @@ bool java_bytecode_languaget::parse(
   }
+}
+
+void java_bytecode_languaget::do_parse()
```
I don't think the name of the function describes well what it does

romainbrenguier commented on jbmc/src/janalyzer/janalyzer_parse_options.cpp:
```diff
@@ -375,6 +375,24 @@ int janalyzer_parse_optionst::doit()
+  if(cmdline.args.size() == 1)
+  {
+    std::string main_class = cmdline.args[0];
+    std::replace(main_class.begin(), main_class.end(), '/', '.');
```
This could be clearer a bit clearer with an explanation or some good naming.

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -8,6 +8,7 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "java_bytecode_language.h"
 
+#include <fstream>
```
commit message doesn't seem to match its content

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: a1919cc2d72bd294801a8753ff7080dc0b6097a6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/138039620
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 03af61633b767fff5e669338589ec6d99aa53998).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/138078934
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
